File: workers/python/python/tasker_core/registry/resolvers/explicit_mapping.py
"""TAS-93: Explicit mapping resolver (priority 10).

This resolver handles explicitly registered handlers with exact key matching.
It's the highest priority resolver in the default chain.

Example:
    >>> resolver = ExplicitMappingResolver()
    >>> resolver.register("my_handler", MyHandlerClass)
    >>>
    >>> definition = HandlerDefinition(callable="my_handler")
    >>> handler = resolver.resolve(definition)
    >>> assert isinstance(handler, MyHandlerClass)
"""

from __future__ import annotations

import threading
from typing import TYPE_CHECKING, Any

from ..base_resolver import BaseResolver

if TYPE_CHECKING:
    from ..handler_definition import HandlerDefinition


class ExplicitMappingResolver(BaseResolver):
    """Resolver for explicitly registered handlers.

    Priority 10 - checked first in the default chain.

    Supports registering:
    - Handler classes (instantiated on resolve)
    - Handler instances (returned directly)
    - Factory functions (called with config on resolve)

    Thread-safe for concurrent registration and resolution.
    """

    def __init__(self, name: str = "explicit_mapping") -> None:
        """Initialize the resolver.

        Args:
            name: Resolver name for identification.
        """
        self._name = name
        self._handlers: dict[str, Any] = {}
        self._lock = threading.RLock()

    @property
    def name(self) -> str:
        """Return the resolver name."""
        return self._name

    @property
    def priority(self) -> int:
        """Return the resolver priority (10 = highest)."""
        return 10

    def can_resolve(
        self,
        definition: HandlerDefinition,
        _config: dict[str, Any] | None = None,
    ) -> bool:
        """Check if this resolver has the handler registered.

        Args:
            definition: Handler configuration.
            _config: Additional context (unused, part of interface).

        Returns:
            True if handler is registered.
        """
        return definition.callable in self._handlers

    def resolve(
        self,
        definition: HandlerDefinition,
        config: dict[str, Any] | None = None,
    ) -> Any | None:
        """Resolve and instantiate a registered handler.

        Args:
            definition: Handler configuration.
            config: Additional context passed to factories.

        Returns:
            Handler instance or None if not registered.
        """
        entry = self._handlers.get(definition.callable)
        if entry is None:
            return None

        return self._instantiate_handler(entry, definition, config or {})

    def register(self, key: str, handler: Any) -> None:
        """Register a handler.

        Args:
            key: Handler identifier (matched against definition.callable).
            handler: Handler class, instance, or factory function.
        """
        with self._lock:
            self._handlers[key] = handler

    def unregister(self, key: str) -> bool:
        """Unregister a handler.

        Args:
            key: Handler identifier to remove.

        Returns:
            True if handler was removed, False if not found.
        """
        with self._lock:
            if key in self._handlers:
                del self._handlers[key]
                return True
            return False

    def registered_callables(self) -> list[str]:
        """Return all registered handler keys.

        Returns:
            List of registered keys.
        """
        return list(self._handlers.keys())

    def _instantiate_handler(
        self,
        entry: Any,
        definition: HandlerDefinition,
        config: dict[str, Any],
    ) -> Any:
        """Instantiate a handler from the registered entry.

        Args:
            entry: Registered handler (class, instance, or factory).
            definition: Handler configuration.
            config: Additional context.

        Returns:
            Handler instance.
        """
        # If it's a class, instantiate it
        if isinstance(entry, type):
            return self._instantiate_class(entry, definition, config)

        # If it's a callable (factory function), call it
        if callable(entry) and not hasattr(entry, "call"):
            return entry(config)

        # Otherwise assume it's already an instance
        return entry

    def _instantiate_class(
        self,
        handler_class: type,
        definition: HandlerDefinition,
        _config: dict[str, Any],
    ) -> Any | None:
        """Instantiate a handler class.

        Tries instantiation without args first, then with config if needed.

        Args:
            handler_class: Handler class to instantiate.
            definition: Handler configuration.
            _config: Additional context (unused, uses definition.initialization).

        Returns:
            Handler instance or None if instantiation fails.
        """
        # First try: instantiate without arguments (most common case)
        try:
            return handler_class()
        except TypeError:
            # May need config - try with it
            pass
        except Exception:
            # Other error - fail
            return None

        # Second try: instantiate with config
        try:
            return handler_class(config=definition.initialization or {})
        except Exception:
            # Instantiation failed - return None
            return None

File: workers/typescript/src/registry/resolvers/explicit-mapping.ts
/**
 * TAS-93: Explicit mapping resolver (priority 10).
 *
 * Resolves handlers from explicitly registered mappings.
 * This is the highest priority resolver in the default chain.
 *
 * Supports registering:
 * - Handler classes (instantiated on resolve)
 * - Handler instances (returned directly)
 * - Factory functions (called with config on resolve)
 *
 * @example
 * ```typescript
 * const resolver = new ExplicitMappingResolver();
 *
 * // Register a class
 * resolver.register('my_handler', MyHandler);
 *
 * // Register an instance
 * resolver.register('shared_handler', new SharedHandler());
 *
 * // Register a factory
 * resolver.register('configurable_handler', (config) => new ConfigHandler(config));
 *
 * // Resolve
 * const definition: HandlerDefinition = { callable: 'my_handler' };
 * const handler = await resolver.resolve(definition);
 * ```
 */

import type { StepHandler, StepHandlerClass } from '../../handler/base.js';
import type { BaseResolver, ResolverConfig } from '../base-resolver.js';
import type { HandlerDefinition } from '../handler-definition.js';

/**
 * Factory function for creating handlers.
 */
export type HandlerFactory = (config: ResolverConfig) => StepHandler;

/**
 * Entry types that can be registered.
 */
export type HandlerEntry = StepHandlerClass | StepHandler | HandlerFactory;

/**
 * Resolver for explicitly registered handlers.
 *
 * Priority 10 - checked first in the default chain.
 */
export class ExplicitMappingResolver implements BaseResolver {
  readonly name: string;
  readonly priority = 10;

  private handlers: Map<string, HandlerEntry> = new Map();

  /**
   * Create an explicit mapping resolver.
   *
   * @param name - Resolver name (default: 'explicit_mapping')
   */
  constructor(name = 'explicit_mapping') {
    this.name = name;
  }

  /**
   * Check if a handler is registered for this callable.
   *
   * @param definition - Handler definition
   * @param _config - Unused, part of interface
   * @returns True if handler is registered
   */
  canResolve(definition: HandlerDefinition, _config?: ResolverConfig): boolean {
    return this.handlers.has(definition.callable);
  }

  /**
   * Resolve and instantiate a registered handler.
   *
   * @param definition - Handler definition
   * @param config - Configuration passed to factories
   * @returns Handler instance or null if not registered
   */
  async resolve(
    definition: HandlerDefinition,
    config?: ResolverConfig
  ): Promise<StepHandler | null> {
    const entry = this.handlers.get(definition.callable);
    if (!entry) {
      return null;
    }

    return this.instantiateHandler(entry, definition, config ?? {});
  }

  /**
   * Register a handler.
   *
   * @param key - Handler identifier (matched against definition.callable)
   * @param handler - Handler class, instance, or factory function
   */
  register(key: string, handler: HandlerEntry): void {
    this.handlers.set(key, handler);
  }

  /**
   * Unregister a handler.
   *
   * @param key - Handler identifier to remove
   * @returns True if handler was removed, false if not found
   */
  unregister(key: string): boolean {
    return this.handlers.delete(key);
  }

  /**
   * Get all registered callable keys.
   *
   * @returns Array of registered keys
   */
  registeredCallables(): string[] {
    return Array.from(this.handlers.keys());
  }

  /**
   * Instantiate a handler from a registered entry.
   */
  private instantiateHandler(
    entry: HandlerEntry,
    definition: HandlerDefinition,
    config: ResolverConfig
  ): StepHandler | null {
    // Check if it's a class (has prototype and is a function)
    if (this.isHandlerClass(entry)) {
      return this.instantiateClass(entry, definition);
    }

    // Check if it's a factory function (function but not a class)
    if (typeof entry === 'function') {
      try {
        return entry(config);
      } catch (error) {
        console.error(`[ExplicitMappingResolver] Factory failed:`, error);
        return null;
      }
    }

    // Otherwise it's an instance - return directly
    return entry as StepHandler;
  }

  /**
   * Check if entry is a handler class.
   */
  private isHandlerClass(entry: HandlerEntry): entry is StepHandlerClass {
    return typeof entry === 'function' && entry.prototype !== undefined && 'handlerName' in entry;
  }

  /**
   * Instantiate a handler class.
   */
  private instantiateClass(
    handlerClass: StepHandlerClass,
    _definition: HandlerDefinition
  ): StepHandler | null {
    try {
      return new handlerClass();
    } catch (error) {
      console.error(`[ExplicitMappingResolver] Failed to instantiate ${handlerClass.name}:`, error);
      return null;
    }
  }
}
